docs/design/sparse-v2.md

```diff
@@ -0,0 +1,225 @@
+rules are subtly re-ordered, they become canonicalized to a smaller, but
+functionally equivalent form:
+
+```txt
+# Before
+include:foo
+exclude:foo/bar/baz/qux
+include:foo/bar/baz
+exclude:foo/bar
+
+# Canonicalized
+include:foo
+exclude:foo/bar
+```
+
+## Working Copy Map
+
+WARNING: This section is intentionally lacking, more research is needed.
+
+All Sparse Patterns v2 will come equipped with a default no-op mapping.
+
+```rust
+vec![WorkingCopyMap {
+    src_path: RepoPathBuf::root(),
+    dst_path: RepoPathBuf::root(),
+}]
+```
+
+`SparsePatterns` will provide an interface to map client paths into repo paths,
+and vice versa. The WorkingCopy trait will apply this mapping to all snapshot
+and checkout operations, and and jj commands which accept relative paths will
+need to be updated to perform client path -> repo path translations as needed.
+It's not clear at this time _which_ commands will need changing, as some are
+more likely to refer to repo paths rather than client paths.
+
+TODO: Expand this section.
+
+In particular, the path rules for sparse patterns for _always_ be repo paths,
+not client paths. Thus, if the client wants to track "foo" and rename it to
+"subdir/bar", they must `jj sparse set --add foo` and
+`jj client-map set --from foo --to bar`. In other words, the mapping operation can
+be thought of as always _after_ the sparse operation.
+
+New commands will enable editing of the WorkingCopyMap:
+
+- `jj client-map list` will print all mapping pairs.
+- `jj client-map set --from foo --to bar` will add a new mapping to the end of the list.
+- `jj client-map remove --from foo` will remove a specific mapping rule.
+- `jj client-map edit` will pull up a text editor for manual editing.
+
+Like sparse paths, mapping rules are defined to apply in _order_ and on any
+save operation will be modified to a minimal canonical form. Thus,
+`jj client-map set --from "" --to ""` will always completely wipe the map.
+The first matching rule in reverse list order determines how a particular
+folder/file should be named in the client.
+
+For simplicity, we require that all client-maps are _invertible_, that is:
+
+- For any repo path R, there is at most one associated client path C
+- Having no associated client path is fine (that part of the repo will never be materialized)
+- For any client path C, there is at most one associated repo path R
+- Having no associated repo path is fine (all files in such paths will be ignored)
+
+To this end, recursive mappings are considers _destructive_. That is, if "foo"
+is renamed to "bar", but "bar" already exists in the repo, then the mapping rule
+hides the original "bar" directory, making it uneditable in the client.
+
+## Versioning and Storage
+
+Updating the active Sparse Pattern for a particular working copy will now
+take place in two separate steps, one which updates the Op Store, and
+another which updates the actual working copy.
```

Comment:
> (I think it seems better to store the full sparse patterns.)

I think for the sake of avoiding redundancy, it's better to only store the Id even if it makes recovery more difficult. (I think that we shouldn't over-design for the case in which the operation is missing.) As an example...

> This is similar to what @jonathantanmy is working on for missing operations (and maybe for missing trees).

...I think the current plan is to make a commit on top of the latest operation and then trigger a snapshot of the working copy, so in theory, we're not using any redundant information like the tree of the working copy (since we can use the tree of the commit in this case, although in practice, I'm not sure what we're using).